Lib/test/test_asyncio/test_queues.py

```diff
@@ -522,5 +522,221 @@ class PriorityQueueJoinTests(_QueueJoinTestMixin, unittest.IsolatedAsyncioTestCa
     q_class = asyncio.PriorityQueue
 
 
+class _QueueShutdownTestMixin:
+    q_class = None
+
+    @staticmethod
+    async def _ensure_started(task):
+        # Explicitly start (if not already) task
+
+        async def swallow(x):
+            try:
+                return await x
+            except Exception:
+                pass
+
+        try:
+            await asyncio.wait_for(asyncio.shield(swallow(task)), 0.01)
+        except TimeoutError:
+            pass
+
+    def assertRaisesShutdown(self, msg="Didn't appear to shut-down queue"):
+        return self.assertRaises(asyncio.QueueShutDown, msg=msg)
+
+    async def test_format(self):
+        q = self.q_class()
+        q._is_shutdown = True
+        self.assertEqual(q._format(), 'maxsize=0 shutdown')
+
+    async def test_shutdown_empty(self):
+        # Test shutting down an empty queue
+
+        # Setup empty queue, and join() and get() tasks
+        q = self.q_class()
+        loop = asyncio.get_running_loop()
+        join_task = loop.create_task(q.join())
+        get_task = loop.create_task(q.get())
+        await self._ensure_started(get_task)  # want pending before shutdown
+
+        # Perform shut-down
+        q.shutdown(immediate=False)  # unfinished tasks: 0 -> 0
+
+        self.assertEqual(q.qsize(), 0)
+
+        # Ensure join() task has successfully finished
+        await self._ensure_started(join_task)
+        self.assertTrue(join_task.done())
+        await join_task
+
+        # Ensure get() task is finished, and raised ShutDown
+        self.assertTrue(get_task.done())
+        with self.assertRaisesShutdown():
+            await get_task
+
+        # Ensure put() and get() raise ShutDown
+        with self.assertRaisesShutdown():
+            await q.put("data")
+        with self.assertRaisesShutdown():
+            q.put_nowait("data")
+
+        with self.assertRaisesShutdown():
+            await q.get()
+        with self.assertRaisesShutdown():
+            q.get_nowait()
+
+    async def test_shutdown_nonempty(self):
+        # Test shutting down a non-empty queue
+
+        # Setup full queue with 1 item, and join() and put() tasks
+        q = self.q_class(maxsize=1)
+        loop = asyncio.get_running_loop()
+
+        q.put_nowait("data")
```

Comment:
Why not just `await q.put("data")`?

Reply:
In this scenario the two are equivalent (the queue is not empty, there are no other tasks scheduled), so I thought `put_nowait` was more explicit